libgalois/src/analytics/bfs/bfs.cpp

```diff
@@ -431,33 +474,13 @@ BfsImpl(
 
   // TODO(lhc): due to lack of in-edge iteration, manually creates a transposed graph
   const katana::GraphTopology& topology = pg->topology();
-  katana::LargeArray<uint32_t> node_data;
-  bool use_block = false;
-  if (use_block) {
-    node_data.allocateBlocked(topology.num_nodes());
-  } else {
-    node_data.allocateInterleaved(topology.num_nodes());
-  }
   auto transpose_graph = katana::CreateTransposeGraphTopology(topology);
 
   katana::do_all(katana::iterate(graph.begin(), graph.end()), [&](auto n) {
     graph.GetData<BfsNodeParent>(n) = BfsImplementation::kDistanceInfinity;
-    node_data[n] = BfsImplementation::kDistanceInfinity;
   });
 
-  katana::StatTimer execTime("BFS");
-  execTime.start();
-  RunAlgo<true>(
-      algo, &graph, pg, *(transpose_graph.value().get()), &node_data, source);
-  execTime.stop();
-
-  if (algo.algorithm() == BfsPlan::kAsynchronous ||
-      algo.algorithm() == BfsPlan::kSynchronousDirectOpt) {
-    katana::do_all(
-        katana::iterate(graph.begin(), graph.end()), [&](auto& node) {
-          graph.GetData<BfsNodeParent>(node) = node_data[node];
-        });
-  }
+  RunAlgo<true>(algo, &graph, pg, *(transpose_graph.value().get()), source);
```

Comment:
I suggest having `InitAlgo` and `FiniAlgo` or `FinalizeAlgo` and you can move code before and after exec timers within `RunAlgo` to those functions with a switch statement.

Reply:
That sounds great but I cannot find clearer ways to declare node data having different type for each algorithm. For example, asynchronous one requires struct type, but direction optimization uses int type. I am still thinking about this.

Reply:
Oh ok. It's not critical.